backend/core.go

```diff
@@ -0,0 +1,40 @@
+// Copyright Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package backend
+
+import (
+	"bytes"
+	"strings"
+	"net/http"
+)
+
+func buildSourceOrigin(host string) string {
+	var sourceOrigin bytes.Buffer
+	if strings.HasPrefix(host, "localhost") {
+		sourceOrigin.WriteString("http://")
+	} else {
+		sourceOrigin.WriteString("https://")
+	}
+	sourceOrigin.WriteString(host)
+	return sourceOrigin.String()
+}
+
+func isFormPostRequest(method string, w http.ResponseWriter,) bool {
```

Comment:
nit: I don't like this method. It does two things (write result code and check for post). At least fix the name to make this more clear. 

A better approach would be to create form handler, e.g. (rough sketch):

```
func handlePost(w http.ResponseWriter, r *http.Request, postHandler func(http.ResponseWriter, *http.Request)) {
    if method != "POST" {
 	http.Error(w, "post only", http.StatusMethodNotAllowed)
 	return;
   }
  postHandler(w, r)
}
```